File: GSTFlow.Rules/Library.fs
namespace GSTFlow.Rules

open System
open GSTFlow.Core

open GSTFlow.Core.Verification

// Raw representation for JSON parsing
type RawParty = {
    Gstin: string
    StateCode: string
    IsSez: bool option
}

type RawInvoiceItem = {
    Hsn: string
    TaxableValue: decimal
    GstRate: decimal
    CessRate: decimal option
    Tax: TaxAmount
}

type RawInvoice = {
    DocumentType: string option
    InvoiceNumber: string
    InvoiceDate: string
    PlaceOfSupply: string option
    OriginalInvoiceNumber: string option
    OriginalInvoiceDate: string option
    Irn: string option
    ReverseCharge: string option
    Seller: RawParty
    Buyer: RawParty option
    Items: RawInvoiceItem list
}



type CompilationResult = {
    IR: GSTCanonicalIR option
    Envelope: VerdictEnvelope
}

module Compiler =
    
    let private createRule outcome id msg =
        { Metadata = { RuleId = id; Category = "GST"; EffectiveFrom = None; EffectiveUntil = None; Reference = None; Confidence = Exact; MessageKey = msg }
          Outcome = outcome
          Evidence = [ { Path = id; Kind = Derived; Value = Some msg; Provenance = Some "Compiler" } ]
          Parameters = Map.empty }
          
    let private failRule id msg = createRule Fail id msg
    let private warnRule id msg = createRule Warning id msg
    let private unknownRule id msg = createRule Unknown id msg

    let validStateCodes = 
        Set.ofList ([ for i in 1..38 -> if i < 10 then "0" + string i else string i ] @ [ "96"; "97"; "99" ])

    let validRateSlabs = 
        Set.ofList [ 0m; 0.1m; 0.25m; 1.5m; 3m; 5m; 12m; 18m; 28m ]

    let isValidHsn (hsn: string) =
        System.Text.RegularExpressions.Regex.IsMatch(hsn, @"^(\d{4}|\d{6}|\d{8})$")

    let private validateParty (role: string) (raw: RawParty) =
        match GSTIN.create raw.Gstin with
        | Ok g ->
            if raw.Gstin.Substring(0, 2) <> raw.StateCode then
                Error (failRule "GSTIN_STATE_MATCH" (role + " StateCode '" + raw.StateCode + "' does not match GSTIN prefix '" + raw.Gstin.Substring(0, 2) + "'"))
            else
                let isSez = match raw.IsSez with Some x -> x | None -> false
                Ok { Party.Gstin = g; Party.StateCode = raw.StateCode; Party.IsSez = isSez }
        | Error e -> Error (failRule "GSTIN_FORMAT" (role + " GSTIN '" + raw.Gstin + "' is invalid: " + e))


    let private validateItem (isInterstate: bool option) (isDocumentRcm: bool) (item: RawInvoiceItem) =
        let mutable violations = []

        if not (isValidHsn item.Hsn) then
            violations <- failRule "HSN_FORMAT" ("HSN '" + item.Hsn + "' must be exactly 4, 6, or 8 digits") :: violations
            
        if not (validRateSlabs.Contains item.GstRate) then
            violations <- failRule "RATE_SLAB" ("GST Rate " + string item.GstRate + " is not a valid Indian slab (0, 0.1, 0.25, 1.5, 3, 5, 12, 18, 28)") :: violations

        let GST_ROUNDING_TOLERANCE = 1.0m // Explicit legally reviewed policy: max combined drift per item

        let expectedTax = Math.Round(item.TaxableValue * (item.GstRate / 100m), 2)
        
        if isDocumentRcm then
            if item.Tax.Igst > 0m || item.Tax.Cgst > 0m || item.Tax.Sgst > 0m then
                violations <- failRule "RCM_CONSISTENCY" "RCM flag/tax amount consistency: RCM flag is Y but tax amounts are present" :: violations
        else
            match isInterstate with
            | Some true ->
                if item.Tax.Cgst > 0m || item.Tax.Sgst > 0m then
                    violations <- failRule "IGST_CGST_LAW" "Interstate supply cannot have CGST or SGST" :: violations
                
                if Math.Abs(item.Tax.Igst - expectedTax) > GST_ROUNDING_TOLERANCE then
                    violations <- failRule "TAX_AMOUNT" ("Expected IGST approx " + string expectedTax + " but got " + string item.Tax.Igst + " (failed item math)") :: violations
            | Some false ->
                if item.Tax.Igst > 0m then
                    violations <- failRule "IGST_CGST_LAW" "Intrastate supply cannot have IGST" :: violations
                
                let expectedSplit = Math.Round(expectedTax / 2m, 2)
                let cDiff = Math.Abs(item.Tax.Cgst - expectedSplit)
                let sDiff = Math.Abs(item.Tax.Sgst - expectedSplit)
                if cDiff + sDiff > GST_ROUNDING_TOLERANCE then
                    violations <- failRule "TAX_AMOUNT" ("Expected CGST/SGST approx " + string expectedSplit + " but got C:" + string item.Tax.Cgst + " S:" + string item.Tax.Sgst) :: violations
            | None ->
                // Math checks only, no law checks
                let totalTax = item.Tax.Igst + item.Tax.Cgst + item.Tax.Sgst
                if Math.Abs(totalTax - expectedTax) > GST_ROUNDING_TOLERANCE then
                    violations <- failRule "TAX_AMOUNT" ("Expected total tax approx " + string expectedTax + " but got " + string totalTax) :: violations
                
        match item.CessRate, item.Tax.Cess with
        | Some crate, Some cval ->
            if not isDocumentRcm then
                let expectedCess = Math.Round(item.TaxableValue * (crate / 100m), 2)
                if Math.Abs(cval - expectedCess) > GST_ROUNDING_TOLERANCE then
                    violations <- failRule "CESS_ARITHMETIC" ("Expected Cess approx " + string expectedCess + " but got " + string cval) :: violations
        | None, Some cval when cval > 0m ->
            if not isDocumentRcm then
                violations <- failRule "CESS_ARITHMETIC" "Cess amount provided but no CessRate specified" :: violations
        | Some _, None ->
            if not isDocumentRcm then
                violations <- failRule "CESS_ARITHMETIC" "CessRate provided but no Cess amount specified" :: violations
        | _ -> ()

        violations

    let compile (raw: RawInvoice) (hash: string) : CompilationResult =
        let mutable violations = []
        
        if String.IsNullOrWhiteSpace(raw.InvoiceNumber) then
            violations <- failRule "INV_SANITY_NO" "InvoiceNumber cannot be empty" :: violations
        
        if String.IsNullOrWhiteSpace(raw.InvoiceDate) then
            violations <- failRule "INV_SANITY_DATE" "InvoiceDate cannot be empty" :: violations
            
        if raw.Items.IsEmpty then
            violations <- failRule "INV_SANITY_ITEMS" "Invoice must contain at least one item" :: violations
            
        for item in raw.Items do
            if item.TaxableValue < 0m then
                violations <- failRule "INV_SANITY_TAXABLE" "Item TaxableValue cannot be negative" :: violations
            if item.GstRate < 0m then
                violations <- failRule "INV_SANITY_RATE" "Item GstRate cannot be negative" :: violations
        
        let sellerRes = validateParty "Seller" raw.Seller
        match sellerRes with
        | Error e -> violations <- e :: violations
        | _ -> ()

        let docType =
            match raw.DocumentType with
            | Some "CRN" -> CRN
            | Some "DBN" -> DBN
            | Some "INV" | None -> INV
            | Some other ->
                violations <- failRule "DOC_TYPE" ("Invalid DocumentType '" + other + "'") :: violations
                INV

        match docType with
        | CRN | DBN ->
            if raw.OriginalInvoiceNumber.IsNone || raw.OriginalInvoiceDate.IsNone then
                violations <- failRule "CDN_ORIGINAL_INV" "Credit/Debit Notes require OriginalInvoiceNumber and OriginalInvoiceDate" :: violations
        | INV -> ()
        
        match raw.Irn with
        | Some irn ->
            if irn.Length <> 64 || not (System.Text.RegularExpressions.Regex.IsMatch(irn, "^[a-fA-F0-9]{64}$")) then
                violations <- failRule "IRN_FORMAT" "IRN must be exactly 64 hexadecimal characters" :: violations
        | None -> ()

        let buyerRes = 
            match raw.Buyer with
            | Some b when String.IsNullOrWhiteSpace(b.Gstin) -> 
                if not (validStateCodes.Contains b.StateCode) then
                    let err = failRule "STATE_CODE" ("Buyer State Code '" + b.StateCode + "' is not in the valid vocabulary (01-38, 97, 99)")
                    violations <- err :: violations
                    Some (Error err)
                else
                    // B2C with known POS
                    let urpGstin = match GSTIN.create "URP" with Ok g -> g | _ -> failwith "URP"
                    Some (Ok { Party.Gstin = urpGstin; Party.StateCode = b.StateCode; Party.IsSez = false })
            | Some b -> 
                match validateParty "Buyer" b with
                | Ok b2 -> Some (Ok b2)
                | Error e -> 
                    violations <- e :: violations
                    Some (Error e)
            | None -> None

        if violations.Length > 0 && violations |> List.exists (fun v -> v.Outcome = Fail) then
            let env = {
                SchemaVersion = "1.0"
                EngineId = "gstflow"
                EngineVersion = "1.0.0"
                RuleSetId = "gstflow-rules"
                RuleSetVersion = "2026.07.10"
                RuleSetDigest = Some "0xMockDigestPendingGovernanceP1_3"
                SubjectType = "gst-invoice"
                SubjectHash = hash
                Results = violations
                OverallOutcome = Fail
            }
            { IR = None; Envelope = env }
        else
            let seller = match sellerRes with Ok s -> s | _ -> failwith "unreachable"
            let buyer = match buyerRes with Some (Ok b) -> Some b | _ -> None
            
            let posEval = PlaceOfSupply.evaluate seller buyer raw.PlaceOfSupply false
            violations <- posEval.Violations @ violations

            let pos =
                match raw.PlaceOfSupply with
                | Some p when not (validStateCodes.Contains p) ->
                    violations <- failRule "PLACE_OF_SUPPLY" ("Invalid PlaceOfSupply '" + p + "'") :: violations
                    p
                | _ -> posEval.EffectivePos

            let isInterstateOpt =
                if pos = "UNKNOWN" then None
                else Some posEval.IsInterstate

            let isInterstate = match isInterstateOpt with Some x -> x | None -> false
            
            let isDocumentRcm =
                match raw.ReverseCharge with
                | Some rc -> rc.ToUpperInvariant() = "Y" || rc.ToUpperInvariant() = "TRUE"
                | None ->
                    match buyer with
                    | Some _ when GSTIN.value seller.Gstin = "URP" -> true // Self Invoice under RCM
                    | _ -> false

            let supplyType = 
                match buyer with
                | Some b when GSTIN.value b.Gstin <> "URP" -> B2B
                | _ -> B2C
                
            let itemViolations = raw.Items |> List.collect (validateItem isInterstateOpt isDocumentRcm)
            violations <- itemViolations @ violations
            
            let envelope = {
                SchemaVersion = "1.0"
                EngineId = "gstflow"
                EngineVersion = "1.0.0"
                RuleSetId = "gstflow-rules"
                RuleSetVersion = "2026.07.10"
                RuleSetDigest = Some "0xMockDigestPendingGovernanceP1_3"
                SubjectType = "gst-invoice"
                SubjectHash = hash
                Results = violations
                OverallOutcome = if violations.IsEmpty then Pass else violations |> List.map (fun v -> v.Outcome) |> List.max
            }

            if violations |> List.exists (fun v -> v.Outcome = Fail) then
                { IR = None; Envelope = envelope }
            else
                let ir = {
                    SourceInvoice = {
                        DocumentType = docType
                        InvoiceNumber = raw.InvoiceNumber
                        InvoiceDate = raw.InvoiceDate
                        OriginalInvoiceNumber = raw.OriginalInvoiceNumber
                        OriginalInvoiceDate = raw.OriginalInvoiceDate
                        Irn = raw.Irn
                        ReverseCharge = isDocumentRcm
                        Seller = seller
                        Buyer = buyer
                        Items = raw.Items |> List.map (fun i -> {
                            Hsn = i.Hsn
                            TaxableValue = i.TaxableValue
                            GstRate = i.GstRate
                            CessRate = i.CessRate
                            Tax = i.Tax
                        })
                    }
                    DerivedSupplyType = supplyType
                    PlaceOfSupply = pos
                    IsInterstate = isInterstate
                }
                { IR = Some ir; Envelope = envelope }

File: playground/src/fable/GSTFlow.Core/Library.ts

import { Record, Union } from "../fable_modules/fable-library-ts.5.6.0/Types.ts";
import { decimal_type, bool_type, class_type, list_type, record_type, option_type, string_type, union_type, TypeInfo } from "../fable_modules/fable-library-ts.5.6.0/Reflection.ts";
import { Option } from "../fable_modules/fable-library-ts.5.6.0/Option.ts";
import { Exception, IComparable, IEquatable } from "../fable_modules/fable-library-ts.5.6.0/Util.ts";
import { FSharpList } from "../fable_modules/fable-library-ts.5.6.0/List.ts";
import { FSharpMap } from "../fable_modules/fable-library-ts.5.6.0/Map.ts";
import { isLetter, isDigit } from "../fable_modules/fable-library-ts.5.6.0/Char.ts";
import { int32 } from "../fable_modules/fable-library-ts.5.6.0/Int32.ts";
import { mapIndexed, sum as sum_1 } from "../fable_modules/fable-library-ts.5.6.0/Array.ts";
import { substring } from "../fable_modules/fable-library-ts.5.6.0/String.ts";
import { create, isMatch } from "../fable_modules/fable-library-ts.5.6.0/RegExp.ts";
import { FSharpResult$2_$union, FSharpResult$2_Error$, FSharpResult$2_Ok } from "../fable_modules/fable-library-ts.5.6.0/Result.ts";
import { decimal } from "../fable_modules/fable-library-ts.5.6.0/Decimal.ts";

export type Verification_RuleConfidence_$union = 
    | Verification_RuleConfidence<0>
    | Verification_RuleConfidence<1>
    | Verification_RuleConfidence<2>

export type Verification_RuleConfidence_$cases = {
    0: ["Exact", []],
    1: ["Derived", []],
    2: ["Guessed", []]
}

export function Verification_RuleConfidence_Exact() {
    return new Verification_RuleConfidence<0>(0, []);
}

export function Verification_RuleConfidence_Derived() {
    return new Verification_RuleConfidence<1>(1, []);
}

export function Verification_RuleConfidence_Guessed() {
    return new Verification_RuleConfidence<2>(2, []);
}

export class Verification_RuleConfidence<Tag extends keyof Verification_RuleConfidence_$cases> extends Union<Tag, Verification_RuleConfidence_$cases[Tag][0]> {
    constructor(tag: Tag, fields: Verification_RuleConfidence_$cases[Tag][1]) {
        super();
        this.tag = tag;
        this.fields = fields;
    }
    readonly tag: Tag;
    readonly fields: Verification_RuleConfidence_$cases[Tag][1];
    cases() {
        return ["Exact", "Derived", "Guessed"];
    }
}

export function Verification_RuleConfidence_$reflection(): TypeInfo {
    return union_type("GSTFlow.Core.Verification.RuleConfidence", [], Verification_RuleConfidence, () => [[], [], []]);
}

export type Verification_RuleOutcome_$union = 
    | Verification_RuleOutcome<0>
    | Verification_RuleOutcome<1>
    | Verification_RuleOutcome<2>
    | Verification_RuleOutcome<3>

export type Verification_RuleOutcome_$cases = {
    0: ["Pass", []],
    1: ["Warning", []],
    2: ["Fail", []],
    3: ["Unknown", []]
}

export function Verification_RuleOutcome_Pass() {
    return new Verification_RuleOutcome<0>(0, []);
}

export function Verification_RuleOutcome_Warning() {
    return new Verification_RuleOutcome<1>(1, []);
}

export function Verification_RuleOutcome_Fail() {
    return new Verification_RuleOutcome<2>(2, []);
}

export function Verification_RuleOutcome_Unknown() {
    return new Verification_RuleOutcome<3>(3, []);
}

export class Verification_RuleOutcome<Tag extends keyof Verification_RuleOutcome_$cases> extends Union<Tag, Verification_RuleOutcome_$cases[Tag][0]> {
    constructor(tag: Tag, fields: Verification_RuleOutcome_$cases[Tag][1]) {
        super();
        this.tag = tag;
        this.fields = fields;
    }
    readonly tag: Tag;
    readonly fields: Verification_RuleOutcome_$cases[Tag][1];
    cases() {
        return ["Pass", "Warning", "Fail", "Unknown"];
    }
}

export function Verification_RuleOutcome_$reflection(): TypeInfo {
    return union_type("GSTFlow.Core.Verification.RuleOutcome", [], Verification_RuleOutcome, () => [[], [], [], []]);
}

export class Verification_RuleMetadata extends Record implements IEquatable<Verification_RuleMetadata>, IComparable<Verification_RuleMetadata> {
    readonly RuleId: string;
    readonly Category: string;
    readonly EffectiveFrom: Option<string>;
    readonly EffectiveUntil: Option<string>;
    readonly Reference: Option<string>;
    readonly Confidence: Verification_RuleConfidence_$union;
    readonly MessageKey: string;
    constructor(RuleId: string, Category: string, EffectiveFrom: Option<string>, EffectiveUntil: Option<string>, Reference: Option<string>, Confidence: Verification_RuleConfidence_$union, MessageKey: string) {
        super();
        this.RuleId = RuleId;
        this.Category = Category;
        this.EffectiveFrom = EffectiveFrom;
        this.EffectiveUntil = EffectiveUntil;
        this.Reference = Reference;
        this.Confidence = Confidence;
        this.MessageKey = MessageKey;
    }
}

export function Verification_RuleMetadata_$reflection(): TypeInfo {
    return record_type("GSTFlow.Core.Verification.RuleMetadata", [], Verification_RuleMetadata, () => [["RuleId", string_type], ["Category", string_type], ["EffectiveFrom", option_type(string_type)], ["EffectiveUntil", option_type(string_type)], ["Reference", option_type(string_type)], ["Confidence", Verification_RuleConfidence_$reflection()], ["MessageKey", string_type]]);
}

export type Verification_EvidenceKind_$union = 
    | Verification_EvidenceKind<0>
    | Verification_EvidenceKind<1>

export type Verification_EvidenceKind_$cases = {
    0: ["Direct", []],
    1: ["Derived", []]
}

export function Verification_EvidenceKind_Direct() {
    return new Verification_EvidenceKind<0>(0, []);
}

export function Verification_EvidenceKind_Derived() {
    return new Verification_EvidenceKind<1>(1, []);
}

export class Verification_EvidenceKind<Tag extends keyof Verification_EvidenceKind_$cases> extends Union<Tag, Verification_EvidenceKind_$cases[Tag][0]> {
    constructor(tag: Tag, fields: Verification_EvidenceKind_$cases[Tag][1]) {
        super();
        this.tag = tag;
        this.fields = fields;
    }
    readonly tag: Tag;
    readonly fields: Verification_EvidenceKind_$cases[Tag][1];
    cases() {
        return ["Direct", "Derived"];
    }
}

export function Verification_EvidenceKind_$reflection(): TypeInfo {
    return union_type("GSTFlow.Core.Verification.EvidenceKind", [], Verification_EvidenceKind, () => [[], []]);
}

export class Verification_Evidence extends Record implements IEquatable<Verification_Evidence>, IComparable<Verification_Evidence> {
    readonly Path: string;
    readonly Kind: Verification_EvidenceKind_$union;
    readonly Value: Option<string>;
    readonly Provenance: Option<string>;
    constructor(Path: string, Kind: Verification_EvidenceKind_$union, Value: Option<string>, Provenance: Option<string>) {
        super();
        this.Path = Path;
        this.Kind = Kind;
        this.Value = Value;
        this.Provenance = Provenance;
    }
}

export function Verification_Evidence_$reflection(): TypeInfo {
    return record_type("GSTFlow.Core.Verification.Evidence", [], Verification_Evidence, () => [["Path", string_type], ["Kind", Verification_EvidenceKind_$reflection()], ["Value", option_type(string_type)], ["Provenance", option_type(string_type)]]);
}

export class Verification_RuleResult extends Record implements IEquatable<Verification_RuleResult>, IComparable<Verification_RuleResult> {
    readonly Metadata: Verification_RuleMetadata;
    readonly Outcome: Verification_RuleOutcome_$union;
    readonly Evidence: FSharpList<Verification_Evidence>;
    readonly Parameters: FSharpMap<string, string>;
    constructor(Metadata: Verification_RuleMetadata, Outcome: Verification_RuleOutcome_$union, Evidence: FSharpList<Verification_Evidence>, Parameters: FSharpMap<string, string>) {
        super();
        this.Metadata = Metadata;
        this.Outcome = Outcome;
        this.Evidence = Evidence;
        this.Parameters = Parameters;
    }
}

export function Verification_RuleResult_$reflection(): TypeInfo {
    return record_type("GSTFlow.Core.Verification.RuleResult", [], Verification_RuleResult, () => [["Metadata", Verification_RuleMetadata_$reflection()], ["Outcome", Verification_RuleOutcome_$reflection()], ["Evidence", list_type(Verification_Evidence_$reflection())], ["Parameters", class_type("Microsoft.FSharp.Collections.FSharpMap`2", [string_type, string_type])]]);
}

export class Verification_VerdictEnvelope extends Record implements IEquatable<Verification_VerdictEnvelope>, IComparable<Verification_VerdictEnvelope> {
    readonly SchemaVersion: string;
    readonly EngineId: string;
    readonly EngineVersion: string;
    readonly RuleSetId: string;
    readonly RuleSetVersion: string;
    readonly SubjectType: string;
    readonly SubjectHash: string;
    readonly Results: FSharpList<Verification_RuleResult>;
    readonly OverallOutcome: Verification_RuleOutcome_$union;
    constructor(SchemaVersion: string, EngineId: string, EngineVersion: string, RuleSetId: string, RuleSetVersion: string, SubjectType: string, SubjectHash: string, Results: FSharpList<Verification_RuleResult>, OverallOutcome: Verification_RuleOutcome_$union) {
        super();
        this.SchemaVersion = SchemaVersion;
        this.EngineId = EngineId;
        this.EngineVersion = EngineVersion;
        this.RuleSetId = RuleSetId;
        this.RuleSetVersion = RuleSetVersion;
        this.SubjectType = SubjectType;
        this.SubjectHash = SubjectHash;
        this.Results = Results;
        this.OverallOutcome = OverallOutcome;
    }
}

export function Verification_VerdictEnvelope_$reflection(): TypeInfo {
    return record_type("GSTFlow.Core.Verification.VerdictEnvelope", [], Verification_VerdictEnvelope, () => [["SchemaVersion", string_type], ["EngineId", string_type], ["EngineVersion", string_type], ["RuleSetId", string_type], ["RuleSetVersion", string_type], ["SubjectType", string_type], ["SubjectHash", string_type], ["Results", list_type(Verification_RuleResult_$reflection())], ["OverallOutcome", Verification_RuleOutcome_$reflection()]]);
}

export function Hash_computeSha256(str: string): string {
    return "hash_not_computed";
}

export function GstinValidation_charToValue(c: string): int32 {
    if (isDigit(c)) {
        return (~~c.charCodeAt(0) - ~~"0".charCodeAt(0)) | 0;
    }
    else if (isLetter(c)) {
        return ((~~c.toLocaleUpperCase().charCodeAt(0) - ~~"A".charCodeAt(0)) + 10) | 0;
    }
    else {
        throw new Exception("Invalid character");
    }
}

export function GstinValidation_valueToChar(v: int32): string {
    if (v < 10) {
        return String.fromCharCode((v + ~~"0".charCodeAt(0)) & 0xFFFF);
    }
    else {
        return String.fromCharCode(((v - 10) + ~~"A".charCodeAt(0)) & 0xFFFF);
    }
}

export function GstinValidation_calculateCheckDigit(gstinWithoutCheck: string): string {
    const remainder: int32 = (sum_1<int32>(mapIndexed<string, int32>((i: int32, c: string): int32 => {
        const product: int32 = (GstinValidation_charToValue(c) * (((i % 2) === 0) ? 1 : 2)) | 0;
        return (~~(product / 36) + (product % 36)) | 0;
    }, gstinWithoutCheck.split(""), Int32Array), {
        GetZero: (): int32 => 0,
        Add: (x: int32, y: int32): int32 => ((x + y) | 0),
    }) % 36) | 0;
    return GstinValidation_valueToChar((remainder === 0) ? 0 : (36 - remainder));
}

export function GstinValidation_isValid(gstin: string): boolean {
    if (gstin.length !== 15) {
        return false;
    }
    else {
        const stateCode: string = substring(gstin, 0, 2);
        if (!isMatch(create(((stateCode === "99") ? true : (stateCode === "97")) ? "^[0-9]{2}[A-Z0-9]{10}[1-9A-Z]{1}[A-Z][0-9A-Z]{1}$" : "^[0-9]{2}[A-Z]{5}[0-9]{4}[A-Z]{1}[1-9A-Z]{1}[Z][0-9A-Z]{1}$"), gstin)) {
            return false;
        }
        else {
            try {
                return GstinValidation_calculateCheckDigit(substring(gstin, 0, 14)) === gstin[14];
            }
            catch (matchValue: any) {
                return false;
            }
        }
    }
}

export class GSTIN extends Union<0, "GSTIN"> {
    constructor(Item: string) {
        super();
        this.tag = 0;
        this.fields = [Item];
    }
    readonly tag: 0;
    readonly fields: [string];
    cases() {
        return ["GSTIN"];
    }
}

export function GSTIN_$reflection(): TypeInfo {
    return union_type("GSTFlow.Core.GSTIN", [], GSTIN, () => [[["Item", string_type]]]);
}

export function GSTINModule_create(str: string): FSharpResult$2_$union<GSTIN, string> {
    if (str === "URP") {
        return FSharpResult$2_Ok<GSTIN, string>(new GSTIN(str));
    }
    else if (GstinValidation_isValid(str)) {
        return FSharpResult$2_Ok<GSTIN, string>(new GSTIN(str));
    }
    else {
        return FSharpResult$2_Error$<GSTIN, string>("Invalid GSTIN format or checksum");
    }
}

export function GSTINModule_value(_arg: GSTIN): string {
    return _arg.fields[0] as string;
}

export class Party extends Record implements IEquatable<Party>, IComparable<Party> {
    readonly Gstin: GSTIN;
    readonly StateCode: string;
    readonly IsSez: boolean;
    constructor(Gstin: GSTIN, StateCode: string, IsSez: boolean) {
        super();
        this.Gstin = Gstin;
        this.StateCode = StateCode;
        this.IsSez = IsSez;
    }
}

export function Party_$reflection(): TypeInfo {
    return record_type("GSTFlow.Core.Party", [], Party, () => [["Gstin", GSTIN_$reflection()], ["StateCode", string_type], ["IsSez", bool_type]]);
}

export type SupplyType_$union = 
    | SupplyType<0>
    | SupplyType<1>

export type SupplyType_$cases = {
    0: ["B2B", []],
    1: ["B2C", []]
}

export function SupplyType_B2B() {
    return new SupplyType<0>(0, []);
}

export function SupplyType_B2C() {
    return new SupplyType<1>(1, []);
}

export class SupplyType<Tag extends keyof SupplyType_$cases> extends Union<Tag, SupplyType_$cases[Tag][0]> {
    constructor(tag: Tag, fields: SupplyType_$cases[Tag][1]) {
        super();
        this.tag = tag;
        this.fields = fields;
    }
    readonly tag: Tag;
    readonly fields: SupplyType_$cases[Tag][1];
    cases() {
        return ["B2B", "B2C"];
    }
}

export function SupplyType_$reflection(): TypeInfo {
    return union_type("GSTFlow.Core.SupplyType", [], SupplyType, () => [[], []]);
}

export class TaxAmount extends Record implements IEquatable<TaxAmount>, IComparable<TaxAmount> {
    readonly Igst: decimal;
    readonly Cgst: decimal;
    readonly Sgst: decimal;
    readonly Cess: Option<decimal>;
    constructor(Igst: decimal, Cgst: decimal, Sgst: decimal, Cess: Option<decimal>) {
        super();
        this.Igst = Igst;
        this.Cgst = Cgst;
        this.Sgst = Sgst;
        this.Cess = Cess;
    }
}

export function TaxAmount_$reflection(): TypeInfo {
    return record_type("GSTFlow.Core.TaxAmount", [], TaxAmount, () => [["Igst", decimal_type], ["Cgst", decimal_type], ["Sgst", decimal_type], ["Cess", option_type(decimal_type)]]);
}

export class InvoiceItem extends Record implements IEquatable<InvoiceItem>, IComparable<InvoiceItem> {
    readonly Hsn: string;
    readonly TaxableValue: decimal;
    readonly GstRate: decimal;
    readonly CessRate: Option<decimal>;
    readonly Tax: TaxAmount;
    constructor(Hsn: string, TaxableValue: decimal, GstRate: decimal, CessRate: Option<decimal>, Tax: TaxAmount) {
        super();
        this.Hsn = Hsn;
        this.TaxableValue = TaxableValue;
        this.GstRate = GstRate;
        this.CessRate = CessRate;
        this.Tax = Tax;
    }
}

export function InvoiceItem_$reflection(): TypeInfo {
    return record_type("GSTFlow.Core.InvoiceItem", [], InvoiceItem, () => [["Hsn", string_type], ["TaxableValue", decimal_type], ["GstRate", decimal_type], ["CessRate", option_type(decimal_type)], ["Tax", TaxAmount_$reflection()]]);
}

export type DocumentType$_$union = 
    | DocumentType$<0>
    | DocumentType$<1>
    | DocumentType$<2>

export type DocumentType$_$cases = {
    0: ["INV", []],
    1: ["CRN", []],
    2: ["DBN", []]
}

export function DocumentType$_INV() {
    return new DocumentType$<0>(0, []);
}

export function DocumentType$_CRN() {
    return new DocumentType$<1>(1, []);
}

export function DocumentType$_DBN() {
    return new DocumentType$<2>(2, []);
}

export class DocumentType$<Tag extends keyof DocumentType$_$cases> extends Union<Tag, DocumentType$_$cases[Tag][0]> {
    constructor(tag: Tag, fields: DocumentType$_$cases[Tag][1]) {
        super();
        this.tag = tag;
        this.fields = fields;
    }
    readonly tag: Tag;
    readonly fields: DocumentType$_$cases[Tag][1];
    cases() {
        return ["INV", "CRN", "DBN"];
    }
}

export function DocumentType$_$reflection(): TypeInfo {
    return union_type("GSTFlow.Core.DocumentType", [], DocumentType$, () => [[], [], []]);
}

export class Invoice extends Record implements IEquatable<Invoice>, IComparable<Invoice> {
    readonly DocumentType: DocumentType$_$union;
    readonly InvoiceNumber: string;
    readonly InvoiceDate: string;
    readonly OriginalInvoiceNumber: Option<string>;
    readonly OriginalInvoiceDate: Option<string>;
    readonly Irn: Option<string>;
    readonly ReverseCharge: boolean;
    readonly Seller: Party;
    readonly Buyer: Option<Party>;
    readonly Items: FSharpList<InvoiceItem>;
    constructor(DocumentType$: DocumentType$_$union, InvoiceNumber: string, InvoiceDate: string, OriginalInvoiceNumber: Option<string>, OriginalInvoiceDate: Option<string>, Irn: Option<string>, ReverseCharge: boolean, Seller: Party, Buyer: Option<Party>, Items: FSharpList<InvoiceItem>) {
        super();
        this.DocumentType = DocumentType$;
        this.InvoiceNumber = InvoiceNumber;
        this.InvoiceDate = InvoiceDate;
        this.OriginalInvoiceNumber = OriginalInvoiceNumber;
        this.OriginalInvoiceDate = OriginalInvoiceDate;
        this.Irn = Irn;
        this.ReverseCharge = ReverseCharge;
        this.Seller = Seller;
        this.Buyer = Buyer;
        this.Items = Items;
    }
}

export function Invoice_$reflection(): TypeInfo {
    return record_type("GSTFlow.Core.Invoice", [], Invoice, () => [["DocumentType", DocumentType$_$reflection()], ["InvoiceNumber", string_type], ["InvoiceDate", string_type], ["OriginalInvoiceNumber", option_type(string_type)], ["OriginalInvoiceDate", option_type(string_type)], ["Irn", option_type(string_type)], ["ReverseCharge", bool_type], ["Seller", Party_$reflection()], ["Buyer", option_type(Party_$reflection())], ["Items", list_type(InvoiceItem_$reflection())]]);
}

export class GSTCanonicalIR extends Record implements IEquatable<GSTCanonicalIR>, IComparable<GSTCanonicalIR> {
    readonly Invoice: Invoice;
    readonly DerivedSupplyType: SupplyType_$union;
    readonly PlaceOfSupply: string;
    readonly IsInterstate: boolean;
    constructor(Invoice: Invoice, DerivedSupplyType: SupplyType_$union, PlaceOfSupply: string, IsInterstate: boolean) {
        super();
        this.Invoice = Invoice;
        this.DerivedSupplyType = DerivedSupplyType;
        this.PlaceOfSupply = PlaceOfSupply;
        this.IsInterstate = IsInterstate;
    }
}

export function GSTCanonicalIR_$reflection(): TypeInfo {
    return record_type("GSTFlow.Core.GSTCanonicalIR", [], GSTCanonicalIR, () => [["Invoice", Invoice_$reflection()], ["DerivedSupplyType", SupplyType_$reflection()], ["PlaceOfSupply", string_type], ["IsInterstate", bool_type]]);
}

